data/src/main/java/com/linkedin/data/schema/annotation/AnnotationCheckResolvedPropertiesVisitor.java

```diff
@@ -49,36 +55,47 @@ public void callbackOnContext(TraverserContext context, DataSchemaTraverse.Order
     }
 
     DataSchema currentSchema = context.getCurrentSchema();
+    RecordDataSchema.Field schemaField = context.getEnclosingField();
+    UnionDataSchema.Member unionMember = context.getEnclosingUnionMember();
 
     ArrayDeque<String> pathToSchema = context.getSchemaPathSpec().clone();
     pathToSchema.addFirst(((NamedDataSchema)context.getTopLevelSchema()).getName());
-    PathSpec pathSpec = new PathSpec(pathToSchema);
-
-    CompatibilityCheckContext compatibilityCheckContext = new CompatibilityCheckContext();
-
-    compatibilityCheckContext.setCurrentDataSchema(currentSchema);
-    compatibilityCheckContext.setSchemaField(context.getEnclosingField());
-    compatibilityCheckContext.setUnionMember(context.getEnclosingUnionMember());
-    compatibilityCheckContext.setPathSpecToSchema(pathSpec);
 
-    // If there is no resolvedProperties but properties, used the properties for annotation check.
-    Map<String, Object> properties;
-    if (context.getEnclosingField() != null)
+    // If current schema is a typeref schema, add TYPEREF_INDICATOR in the pathSpec
+    // to avoid this node and it's child node have the same pathSpec.
+    if (currentSchema instanceof TyperefDataSchema)
     {
-      RecordDataSchema.Field field = context.getEnclosingField();
-      properties = chooseProperties(field.getResolvedProperties(), field.getProperties());
+      context.getSchemaPathSpec().addLast(TYPEREF_INDICATOR);
     }
-    else if (context.getEnclosingUnionMember() != null)
+
+    // Current node is a field of a record schema, get the field's annotation.
+    if (schemaField != null && pathToSchema.getLast().equals(context.getEnclosingField().getName()))
     {
-      // There is no resolvedProperty for unionMember
-      properties = context.getEnclosingUnionMember().getProperties();
+      // Add FIELD_INDICATOR in the pathSpec to differentiate field annotation and field type schema annotation.
+      pathToSchema.addLast(FIELD_INDICATOR);
+      PathSpec pathSpec = new PathSpec(pathToSchema);
+
+      _nodeToResolvedPropertiesMap.put(pathSpec,
+          new ImmutablePair<>(generateCompatibilityCheckContext(schemaField, unionMember, currentSchema, pathSpec),
+              chooseProperties(schemaField.getResolvedProperties(), schemaField.getProperties())));
+      pathToSchema.removeLast();
     }
-    else
+
+    // Current node is a union member, get the union member key's annotation.
```

Comment:
I am wondering if it could be written into a single
if ... else if ... block
instead of two if blocks
Because that way, we will get to know this two logic are meant for output just one thing:
a pathspec to checkerContext map